### ISSUE TITLE
Feature request - more flexible configuration blocks

### ISSUE THREAD

**Issue by SirCmpwn** [i3/i3#2351]
I intend to do this for sway and it would be a cool feature for i3 to have as well: https://github.com/SirCmpwn/sway/issues/547

>It would be nice if we tweaked the implementation of command blocks to work for any command. That is, things like this:
>
>    output HDMI-A-1 {
>        pos 1920,0
>        res 1920x1080
>        bg ~/wallpaper.png
>    }
>
>Would be expanded and run as three seperate commands:
>
>    output HDMI-A-1 pos 1920,0
>    output HDMI-A-1 res 1920x1080
>    output HDMI-A-1 bg ~/wallpaper.png
>
>Existing command blocks would change so that this:
>
>    bar {
>        status_command custom_statusbar
>        position top
>        colors {
>            statusline #ffffff
>            background #323232
>            focused_workspace #32323200 #32323200 #ffffff
>            inactive_workspace #00000000 #32323200 #666666
>            active_workspace #00000000 #32323200 #666666
>        }
>    }
>
>Would be equivalent to this:
>
>    bar status_command custom_statusbar
>    bar position top
>    bar colors statusline #ffffff
>    bar colors background #323232
>    bar colors focused_workspace #32323200 #32323200 #ffffff
>    bar colors inactive_workspace #00000000 #32323200 #666666
>    bar colors active_workspace #00000000 #32323200 #666666
>
>I think this would simplify our implementation of command blocks, simplify the documentation, and allow for more flexibility.

**Comment by stapelberg** [comment]
It’s an intriguing feature, but I don’t see a motivating use case. i3 doesn’t have output configuration built-in, so the new output block wouldn’t make much sense.

Can you elaborate on whether there are other examples you have in mind that would apply to i3 itself?

**Comment by SirCmpwn** [comment]
Outputs of course would not apply to i3, but the feature itself I think is solid regardless. For example, this would be possible:

    bindsym {
        $mod+j move up
        $mod+k move down
        # ...
    }

The currently crappy color syntax might become:

    colors {
        client focused x y z q
        client unfocused x y z q
    }

Or:

    colors {
        client {
            focused x y z q
            unfocused x y z q
        }
        # ...future features
    }

**Comment by SirCmpwn** [comment]
Would also just make for a more flexible and DRY implementation codewise.

**Comment by acrisci** [comment]
I think that could be a good feature.

**Comment by stapelberg** [comment]
Thanks for your patience while I’ve been very busy and took some time to consider this feature.

My gut feeling tells me that this is not a feature I’d like to have in i3. Here are a few thoughts to hopefully provide enough context that you can understand my line of thinking:

* I agree that it’s less typing and less repetition, but i3’s target audience (advanced users and programmers) are expected to know how to leverage their text editor to deal with the repetiton, so that aspect of the proposed change isn’t so appealing to me.
* Subjectively, I feel configuration files become more complicated when using blocks e.g. for bindsym, because you then need to be more aware of the context (i.e. where you are in the file) when making changes/when reading the file.
* In a way, this proposed change divides existing configuration files and new configuration files into a de-facto old and new format, which I think should be avoided unless absolutely necessary. For the v3→v4 transition, we wrote a converter tool and the whole experience was painful enough that I’m not looking to do big changes to the configuration structure again.
 * As a sub point of this: exchanging individual parts of the configuration file would be slightly harder than before, because users might not use a version of i3 which supports the new format. Since especially i3 beginners edit their configuration file while figuring out how to best use i3, I don’t want to make it harder for them than it has to be.

**Comment by SirCmpwn** [comment]
Thanks for the detailed response. I think that's fair, I'll close the issue in sway as well.

**Comment by SirCmpwn** [comment]
This has come up for discussion again on sway and I have to say I'm no longer swayed by your arguments (hehe, *swayed*).

>I agree that it’s less typing and less repetition, but i3’s target audience (advanced users and programmers) are expected to know how to leverage their text editor to deal with the repetiton, so that aspect of the proposed change isn’t so appealing to me.

Advanced users and programmers should appreciate DRY more than anyone else, really.

>Subjectively, I feel configuration files become more complicated when using blocks e.g. for bindsym, because you then need to be more aware of the context (i.e. where you are in the file) when making changes/when reading the file.

This hasn't been a problem for programmers for years, who just use indentation to that effect. Following your earlier argument, they should be comfortable enough using their text editor to hop back to the start of the block and see what it does, if necessary (which I imagine it often won't be).

>In a way, this proposed change divides existing configuration files and new configuration files into a de-facto old and new format, which I think should be avoided unless absolutely necessary. For the v3→v4 transition, we wrote a converter tool and the whole experience was painful enough that I’m not looking to do big changes to the configuration structure again.

This change is backwards compatible with old formats. I don't think forward compatability is a good goal because it prevents you from ever innovating and isn't all that useful.

**Comment by Airblader** [comment]
> This change is backwards compatible with old formats.

I'm not advocating for or against this feature, but if it's backwards compatible – what stops you from implementing it in sway even if it's not in i3?

**Comment by SirCmpwn** [comment]
Nothing, and I think we intend to, but it would be nice if sway and i3 configs were cross-compatible. I also just like pitching sway features to i3 so that we both benefit from each other's improvements.

**Comment by Airblader** [comment]
> but it would be nice if sway and i3 configs were cross-compatible

Why? :-) The motivation of sway was (is) to allow X11 users to switch to Wayland without any headache. What reason would a user have to switch from Wayland to X11?

Compatibility with sway has never been a concern of i3, only the other way around. Given that we don't follow sway development (nor influence it) I don't think this is a goal we should be taking on.

>  I also just like pitching sway features to i3 so that we both benefit from each other's improvements.

And that is, of course, much appreciated. :-)

On the topic – I really have no strong opinion either way. But some thoughts:

* I don't think DRY applies here, because, comparing this to a programming language, we don't criticize language designers for having to repeat `while` and similar keywords everytime they're used. DRY would apply if a user had to configure the same thing multiple times, but they don't need to do that.
* It's also one of i3's selling points that the config is rather KISS (to throw another paradigm out there) as opposed to, say, awesome where you basically write scrips and have to understand a complex syntax. I'm not saying either is better than the other, just that they're different and this feels like a step away from our current concept.
* There's also projects like i3bang (which has sort of died out, but is working just fine) which allow much, much more in terms of keeping things short.

But that said, I'm also not opposed to supporting this if somebody was willing to implement it.

**Comment by SirCmpwn** [comment]
>Why? :-) The motivation of sway was (is) to allow X11 users to switch to Wayland without any headache. What reason would a user have to switch from Wayland to X11?

X isn't dead yet, and won't be for some time. Even I still use it from time to time. I don't want to lock users into Wayland any more than you want to lock them into X11. I've put work into features that should allow users to switch between both as painlessly as possible.

>Compatibility with sway has never been a concern of i3, only the other way around. Given that we don't follow sway development (nor influence it) I don't think this is a goal we should be taking on.

Well, I suppose I can't do anything about that. It would be nice if, for example, when i3 adds a new command if you guys could hop over to sway and double check that it doesn't conflict with one of ours.

>I don't think DRY applies here, because, comparing this to a programming language, we don't criticize language designers for having to repeat while and similar keywords everytime they're used. DRY would apply if a user had to configure the same thing multiple times, but they don't need to do that.
>
>It's also one of i3's selling points that the config is rather KISS (to throw another paradigm out there) as opposed to, say, awesome where you basically write scrips and have to understand a complex syntax. I'm not saying either is better than the other, just that they're different and this feels like a step away from our current concept.

I think this change appeals to both DRY and KISS, but perhaps more to KISS. Right now, blocks in i3 configs are a special case that behaves differently from other commands. It even introduces a special syntax that only applies to specific commands (mode, bar, bar > color), and a new command set that only works within these blocks. Special cases are more complex, not less. The proposed change makes all commands behave consistently. There are no commands that are valid in special cases. All commands can have a block.

**Comment by Airblader** [comment]
> It would be nice if, for example, when i3 adds a new command if you guys could hop over to sway and double check that it doesn't conflict with one of ours.

I did do this occasionally, in particular in the beginning or to give you guys a heads up on the `set_from_resource`. Which is actually a relatively good example of the problem we're facing. Syntax-wise there was no issue with adding it to sway, but functionally it as simply impossible to implement. But while it has no equivalent in Wayland, it was still a valuable feature under X. sway actually got really lucky that we had to add a fallback value for a couple of reasons (none of which were that we considered sway); if this hadn't been the case, sway would've had a problem here (and even now the situation isn't ideal).

»Collisions« in command names, on the other hand, are something I see as worth avoiding (though perhaps I would recommend you to watch the i3 repository for this instead and intervening if there is a collision) in terms of a common goal. Perhaps sway should have had some mechanism to namespace sway-only commands, but hindsight is 20/20. Then again, we never ran into collisions this far and I think it's unlikely that we will.

> It even introduces a special syntax that only applies to specific commands (mode, bar, bar > color), and a new command set that only works within these blocks. Special cases are more complex, not less. The proposed change makes all commands behave consistently. There are no commands that are valid in special cases. All commands can have a block.

To me this is really the best argument in favor of this change that I've heard so far.

**Comment by SirCmpwn** [comment]
>I did do this occasionally, in particular in the beginning or to give you guys a heads up on the set_from_resource. Which is actually a relatively good example of the problem we're facing. Syntax-wise there was no issue with adding it to sway, but functionally it as simply impossible to implement. But while it has no equivalent in Wayland, it was still a valuable feature under X. sway actually got really lucky that we had to add a fallback value for a couple of reasons (none of which were that we considered sway); if this hadn't been the case, sway would've had a problem here (and even now the situation isn't ideal).

Well, not necessarily. I have always had the position that sway will not implement things that just don't make sense on Wayland, and I have actually in a couple of places agreed to break backwards compatability in exchange for saner behavior. The goal isn't to be perfect, but in the broader strokes we should be as compatible as possible.

**Comment by stapelberg** [comment]
>> […special cases argument…]
>
> To me this is really the best argument in favor of this change that I've heard so far.

While I agree in principle that reducing special cases is a good idea, keep in mind that there is a significant migration cost for any change we do to the i3 configuration file. This cost increases with the scope of the change, and I can’t imagine a bigger scope than such a basic change in syntax. Such a change is hard to communicate to users; see also questions we get about command escaping (which is a syntax specific subtlety that _is_ documented, but doesn’t jump at people when they read our documentation).

I also hear the point about forward/backward-compatibility and it hindering us to innovate. That is true, but with i3’s current level of maturity, I value broad configuration file compatibility (i.e. the chance that a random i3 tutorial you find on the internet or in a computer magazine (!) is usable with the version on your computer) over innovation in configuration file syntax.

With regards to the DRY principle and other arguments in the same spirit, I’d like to remark that the i3 config file is — the way I perceive it — not something you edit every day (I would be very curious about hard numbers on how often people adjust their i3 config file, but I think such numbers are hard to get). Hence, I think clarity trumps DRY in this case.

In summary, I’m still not convinced :).

**Comment by Airblader** [comment]
We wouldn't need a migration if the change is completely compatible. And that's the only scenario where I'd favor this change. We definitely shouldn't break configs.

**Comment by stapelberg** [comment]
While it doesn’t need to be an enforced migration, it still is a kind of migration. People will have differing levels of motivation to change their config file to use the new syntax. Let’s not get too hung up on terminology, though :).

**Comment by SirCmpwn** [comment]
No, let's. Existing configs will work as-is on the new i3 and no changes are required to them. What kind of migration is that? A random i3 tutorial you found on the internet will still work because the change is backwards compatible. There is no migration cost here.

**Comment by Airblader** [comment]
Yeah, I'm also confused as to why a backwards-compatible change would be considered a migration. In that sense adding a new config directive should be considered a migration (?)

**Comment by stapelberg** [comment]
> A random i3 tutorial you found on the internet will still work because the change is backwards compatible.

Pre-block tutorials will work. Post-block tutorials might only work with new enough versions. Imagine a user finds a post-block tutorial and compares the config file with their own config file. It looks different due to all the blocks. Now the user has to make a decision:

1. Ignore the tutorial and continue life.
2. Change the existing config file to use blocks (in my terminology: “migrate” the config file from old to new). Might require an i3 upgrade.
3. Understand how the “migration”/change from old to new works and backport the tutorial.

> In that sense adding a new config directive should be considered a migration (?)

Additions are a trivial special case of a migration, because there is nothing to change.

Changing an existing config directive I would also call migration, with the difference being in scope (single directive vs. syntax).

I hope this made it a bit clearer what I meant. Every change comes with a mental cost that we’re imposing on our users. In this case, I don’t think the benefits outweigh that cost.

**Comment by Airblader** [comment]
> Post-block tutorials might only work with new enough versions.

And this definitely will happen because it already happens with new config directives and commands. So it's not really a new situation.

Of course the cost–benefit consideration might be different because new features add a much bigger benefit than being able to achieve the same thing in a different way.

**Comment by SirCmpwn** [comment]
Also consider that many i3 users *enjoy* tweaking their configs and find it well worth their time to learn how it works given the important role it has in their workflow.
